### DISCUSSION TITLE
Typing: `on_event` method is missing generic arguments

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/starlette#1523] [Potential Issue] [answered]
Pyright does not like the `on_event` method in strict mode:

```py
from starlette.applications import Starlette

app = Starlette()

@app.on_event('startup')  # E: Untyped function decorator obscures type of function; ignoring decorator
async def on_startup() -> None:
    pass
```

This can be fixed by adding generic arguments to the `on_event` definition:

```py
F = TypeVar('F', bound=typing.Callable[..., Any])

def on_event(self, event_type: str) -> typing.Callable[[F], F]:
    return self.router.on_event(event_type)
```

I know the usage of these decorators is discouraged but it is necessary for some use cases: https://github.com/RobertCraigie/prisma-client-py/issues/290

**Comment by adriangb** [comment]
Could you provide a context manager? This reduces the boilerplate a bit when used with Starlette.

I don't think the team is too keen on adding more typing info, but @tomchristie can give the final word on it

**Reply by RobertCraigie** [reply]
Sorry, I'm not sure I understand. How could a context manager be used in this situation?

**Reply by adriangb** [reply]
Maybe something like:

```python
client = prisma.Client(...)

Starlette(
    lifespan=prisma.starlette.Lifespan(client)
)
```

Where pristma.starlette.Lifespan is a context manager that just discards the first argument (Starlette forces you to accept an instance of itself) and then calls prisma.Client.__aenter__/__aexit__.

And by making prisma.Client an async context manager, users have an existing lifespan they can use it in there without the wrapper.

**Reply by RobertCraigie** [reply]
Ah I see , I didn't know about lifespan, thank you.

This would work but wouldn't it then be awkward for users / other frameworks to define their own startup / shutdown events as there can only be one lifespan?

**Reply by adriangb** [reply]
Yep hence why I would also make Client itself an async context manager so that if they are defining their own lifespan they can use it directly. The prisma.starlette.Lifespan would just be to reduce boilerplate in the simple case where they are not using their own lifespan. Personally I don't think it's worth it but if you want a little boiler plate solution for users then that is one way to do it

**Comment by lovelydinosaur** [comment]
> Pyright does not like the on_event method in strict mode

Rather than use the decorator, use the plain ol' `on_startup`/`on_shutdown` keyword arguments when instantiating the app.

https://www.starlette.io/applications/

We don't document the decorator style anymore, and there's a code comment recommending against it.

We'll move it to a more formal deprecation at some point.

**Reply by RobertCraigie** [reply]
This would make it annoying for consumers to use the helper functions I plan on adding to my library.

```py
from starlette.application import Starlette
from prisma.ext.starlette import prisma_startup, prisma_shutdown

starlette = Starlette(
    on_startup=[prisma_startup],
    on_shutdown=[prisma_shutdown],
)
```

It also makes it error prone. Consider the situation that in the future these helper startup / shutdown methods only work if the app is configured in a certain way, anyone using the helper methods will then have to know that they need to update their code instead of it just being handled for them like this solution would support:

```py
from starlette.application import Starlette
from prisma.ext.starlette import register_prisma

starlette = Starlette()
register_prisma(starlette)
```

Was this use case considered when deciding to discourage using the decorator style?

**Comment by Kludex** [comment, marked as answer]
The decorators are deprecated on the latest Starlette release.
